packages/eslint-plugin/src/rules/prefer-at.ts

```diff
@@ -0,0 +1,177 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'tsutils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'prefer-at',
+  meta: {
+    type: 'suggestion',
+    fixable: 'code',
+    docs: {
+      description:
+        'Enforce the use of `array.at(-1)` instead of `array[array.length - 1]`',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    messages: {
+      preferAt:
+        'Expected a `{{name}}.at(-1)` instead of `{{name}}[{{name}}.length - 1]`.',
+    },
+    schema: [
+      {
+        oneOf: [
+          {
+            type: 'object',
+            properties: {
+              ignoreFunctions: {
+                type: 'boolean',
+              },
+            },
+            additionalProperties: false,
+          },
+        ],
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      ignoreFunctions: false,
+    },
+  ],
+  create(context, [options]) {
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+    const sourceCode = context.getSourceCode();
+
+    function getName(node: TSESTree.Node): string | undefined {
+      switch (node.type) {
+        case AST_NODE_TYPES.Identifier:
+          return node.name;
+        case AST_NODE_TYPES.MemberExpression:
+          return getName(node.property);
+        default:
+          return undefined;
```

Comment:
[Testing] Still missing some coverage here. See https://app.codecov.io/gh/typescript-eslint/typescript-eslint/pull/6411/blob/packages/eslint-plugin/src/rules/prefer-at.ts for three lines that are yellow (partial miss) or red (full miss).